### PR TITLE
8137018: [JVMCI] Encapsulate new Thread fields for JVMCI

### PR DESCRIPTION
**Pull request by tkrodriguez**
This evacuates all JVMCI related methods and fields into a separately declared struct.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8137018](https://bugs.openjdk.java.net/browse/JDK-8137018): [JVMCI] Encapsulate new Thread fields for JVMCI

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 9af54d4b9a51265daf2e565c6593593c01e11256
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 9af54d4b9a51265daf2e565c6593593c01e11256
 * [Doug Simon](https://openjdk.java.net/census#dnsimon) (@dougxc - Committer) ⚠️ Review applies to 472073a726fc0b7e3f1e2da7ed23c7a1ea8f2a49

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5339/head:pull/5339` \
`$ git checkout pull/5339`

Update a local copy of the PR: \
`$ git checkout pull/5339` \
`$ git pull https://git.openjdk.java.net/jdk pull/5339/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5339`

View PR using the GUI difftool: \
`$ git pr show -t 5339`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5339.diff">https://git.openjdk.java.net/jdk/pull/5339.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/jvmci/jvmci.hpp [hunk 1]
```diff
@@ -163,6 +163,115 @@ class JVMCI : public AllStatic {
...
+  static void free_counters(JavaThread* thread);
+
+  static void accumulate_counters(JavaThread* thread);
+
+  static void init_counters();
+
+  static void free_counters();
+};
+
+
```
Comment (coleenp) [inline src/hotspot/share/jvmci/jvmci.hpp:183]: There seems to be a crazy amount of whitespace here.

**Review by coleenp — APPROVED** [review]
Looks awesome, Thanks!

File: src/hotspot/share/jvmci/jvmci.cpp [hunk 2]
```diff
@@ -290,3 +292,160 @@ void JVMCI::event3(const char* format, ...) LOG_TRACE(3)
...
+  if (JVMCICounterSize > 0) {
+    if (jvmci_counters_include(thread)) {
+      for (int i = 0; i < JVMCICounterSize; i++) {
+        _jvmci_old_thread_counters[i] += thread->jvmci()._jvmci_counters[i];
+      }
+    }
+  }
+}
+
+
```
Comment (dholmes-ora) [inline src/hotspot/share/jvmci/jvmci.cpp:405]: Nit: there are a few double-blank lines between definitions when one is normal.

File: src/hotspot/share/jvmci/jvmci.hpp [hunk 3]
```diff
@@ -163,6 +163,115 @@ class JVMCI : public AllStatic {
...
+ private:
+  friend class JavaThread;
+  friend class JVMCI;
+  friend class JVMCIVMStructs;
+
+  // The _pending_* fields below are used to communicate extra information
+  // from an uncommon trap in JVMCI compiled code to the uncommon trap handler.
+
+  // Communicates the DeoptReason and DeoptAction of the uncommon trap
+  int       _pending_deoptimization;
```
Comment (dholmes-ora) [inline src/hotspot/share/jvmci/jvmci.hpp:198]: Nit: Why the extra large alignment spacing of all the declarations? (I'm not a fan of such alignment as it is too hard to maintain - and too hard to type in the first place!)

File: src/hotspot/share/jvmci/jvmci.hpp [hunk 4]
```diff
@@ -163,6 +163,115 @@ class JVMCI : public AllStatic {
...
+  void set_pending_failed_speculation(jlong failed_speculation) { _pending_failed_speculation = failed_speculation; }
+  void set_pending_transfer_to_interpreter(bool b) { _pending_transfer_to_interpreter = b; }
+  void set_jvmci_alternate_call_target(address a) { assert(_union._alternate_call_target == NULL, "must be"); _union._alternate_call_target = a; }
+  void set_jvmci_implicit_exception_pc(address a) { assert(_union._implicit_exception_pc == NULL, "must be"); _union._implicit_exception_pc = a; }
+  void set_in_retryable_allocation(bool b)        { _in_retryable_allocation = b; }
+  bool in_retryable_allocation() const            { return _in_retryable_allocation; }
+
+  void set_jvmci_reserved_oop0(oop value) {
+    _jvmci_reserved_oop0 = value;
+  }
```
Comment (dholmes-ora) [inline src/hotspot/share/jvmci/jvmci.hpp:249]: Nit: why is this and following definitions multi-line when the preceding ones (of similar size) are single line?

**Review by dholmes-ora — APPROVED** [review]
Nice refactoring!

A few style nits but nothing of consequence.

Thanks,
David

File: src/hotspot/share/jvmci/jvmci.hpp (hunk 3 above)
Comment (dougxc) [reply src/hotspot/share/jvmci/jvmci.hpp:198]: That probably comes from a time before we nicely commented each JVMCI field ;-)
I agree that there's no need for the alignment now.

**Review by dougxc — APPROVED** [review]

File: src/hotspot/share/jvmci/jvmci.hpp (hunk 4 above)
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmci.hpp]: I've flattened them out.

File: src/hotspot/share/jvmci/jvmci.hpp (hunk 3 above)
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmci.hpp]: I removed the extra indenting.

File: src/hotspot/share/jvmci/jvmci.cpp (hunk 2 above)
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmci.cpp]: fixed.

File: src/hotspot/share/jvmci/jvmci.hpp (hunk 1 above)
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmci.hpp:188]: I'd meant to go back and pretty this up.  I was just dumping things in there as I migrated them from elsewhere.  I've grouped them with some short comments and tried to clean it up.  I also fixed some visibility issues and removed the use of friend for JavaThread.

**Comment by tkrodriguez** [comment]
I pushed a commit that address all the comments and cleans up the header a bit.  Testing of this commit was clean.  Hopefully I haven't introduced any new things which should be address.

**Review by dholmes-ora — APPROVED** [review]
Looks fine. Thanks for the updates.

David

**Review by coleenp — APPROVED** [review]
Update looks fine.

**Pull request closed without merging**